src/main/java/teammates/ui/webapi/RequestTraceFilter.java

```diff
@@ -0,0 +1,88 @@
+package teammates.ui.webapi;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.http.HttpStatus;
+
+import teammates.common.exception.TeammatesException;
+import teammates.common.util.Logger;
+import teammates.common.util.RequestTracer;
+
+/**
+ * Extracts trace ID of HTTP requests.
+ */
+public class RequestTraceFilter implements Filter {
+
+    private static final Logger log = Logger.getLogger();
+
+    @Override
+    public void init(FilterConfig filterConfig) {
+        // nothing to do
+    }
+
+    @Override
+    public void doFilter(ServletRequest req, ServletResponse resp, FilterChain chain) throws IOException {
+        HttpServletRequest request = (HttpServletRequest) req;
+        HttpServletResponse response = (HttpServletResponse) resp;
+
+        // The header X-AppEngine-QueueName cannot be spoofed as GAE will strip any user-sent X-AppEngine-QueueName headers.
+        // Reference: https://cloud.google.com/appengine/docs/standard/java/taskqueue/push/creating-handlers#reading_request_headers
+        boolean isRequestFromAppEngineQueue = request.getHeader("X-AppEngine-QueueName") != null;
+
+        // GAE will terminate an instance if any request exceeds 10 minutes.
+        // For GAE-invoked requests, we set the limit here minus a small grace period of 5 seconds
+        // to ensure that the 10 minutes limit will not be exceeded.
+        // For user-invoked requests, we keep the time limit at 1 minute (as how it was
+        // in the previous GAE runtime environment) in order to not let user wait for excessively long,
+        // as well as a reminder for us to keep optimizing our API response time.
+        int timeoutInSeconds = isRequestFromAppEngineQueue ? 10 * 60 - 5 : 60;
```

Comment:
Is the grace period of `5 seconds` sufficient? The instance might still get terminated while we do the teardown steps.

Reply:
I don't have data to back up but I don't believe tearing down will take that long.